### PR TITLE
RSDK-6713 RestartModule endpoint + CLI command

### PR DESCRIPTION
**Pull request by abe-winter**
## What changed
- add `localRobot.RestartModule` (internally this calls `updateResources` with a faked config.Diff)
- cli: `viam module reload` command. currently only works with `--restart-only` flag
- I didn't add many tests because this is mostly wrapper code, but if a reviewer thinks we need them, I can add something similar to [TestModuleReloading here](https://github.com/viamrobotics/rdk/blob/main/module/modmanager/manager_test.go#L395)

Note: this is not the full 'reload' feature (that's coming in next PR). This is just the inner 'restart' feature.
## Questions for reviewer
- is CLI supposed to be doing grpc debug logging? is it wrong to pass `logging.Global()` in module_build.go
## Checklist
- [x] wait for linked API PR https://github.com/viamrobotics/api/pull/478, then remove `replace` directive in go.mod

### CODE REVIEW

File: cli/module_build.go
Comment (zaporter-work) [inline cli/module_build.go:457]: I think we should be able to use the cached credentials / setup some pipieline with app. I don't think we need to accept creds here

Comment (zaporter-work) [reply cli/module_build.go:457]: Brainstorming: Look at connect code samples page API requests to see if we can mimic that

File: robot/impl/local_robot.go
Comment (zaporter-work) [inline robot/impl/local_robot.go:1223]: I like this approach (instead of updating the module from underneath the rest of the robot).

**Review by zaporter-work — COMMENTED** [review]
First pass. Looking like a good approach

File: cli/module_build.go
Comment (abe-winter) [reply cli/module_build.go:44]: yes -- added it because I wanted there to be 'just one thing', but I'm not sure this is the thing

imo if we're multiplexing, we should set up a prefix so it's canonical + hard to break? like `kv:uuid:base64`?

File: robot/impl/local_robot.go
Comment (abe-winter) [reply robot/impl/local_robot.go:1223]: I initially tried calling the module's `Reconfigure` directly, but it removed the resources + then didn't add them back -- `updateResources` did the right thing, and it respects the existing mutexes

File: cli/module_build.go
Comment (abe-winter) [reply cli/module_build.go:457]: I'm guessing the cloud app RobotService is not a good option? https://github.com/viamrobotics/api/blob/main/proto/viam/app/v1/robot.proto#L13-L25
like we don't want it to be a passthrough for the full functionality of the device

Comment (abe-winter) [reply cli/module_build.go:457]: ah you mean the keys list on `/machine/xxx/connect/api-keys`? that makes sense to me

are there downsides to this? use wrong key by accident?

Comment (zaporter-work) [reply cli/module_build.go:457]: To my knowledge, all robots are auto-generated an APIKey such that resource_type = robot and resource_id = {robot you want to connect to}.  I think there is no harm in finding that key and using that key to auth to the robot

(the user can delete that key, but we can almost-certainly treat that as an edge case, a few parts of our app complain if there are no api keys for a robot)

Comment (zaporter-work) [reply cli/module_build.go:457]: It seems like significantly better UX than making the user go to app.viam (and login via the same auth that they signed into the cli with, and go manually copy-paste api keys into their terminal history)

File: robot/impl/local_robot.go
Comment (benjirewis) [reply robot/impl/local_robot.go:1223]: I'm worried just calling `updateResources` won't work in some circumstances, but let's chat offline.

File: cli/app.go
Comment (benjirewis) [inline cli/app.go:21]: Hopefully you won't need this nor `loginFlagKeyIDVal` if you rely on existing auth mechanisms as @zaporter-work suggested.

File: cli/module_build.go
Comment (benjirewis) [reply cli/module_build.go:44]: > I'm going to try viamClient / getapikeys approach, so this auth approach may go away

Talked a bit offline, and I agree with relying on existing auth approach/getting rid of this.

File: cli/module_build.go
Comment (benjirewis) [reply cli/module_build.go:53]: Agree with @zaporter-work: let's keep this in `cli/app.go`. Eventually, you are correct @abe-winter that `cli/app.go` will get bloated. At that point; I think we can break `cli/app.go` into smaller `app.go` files in subdirectories specifying pieces of the CLI interface (this piece could go in `cli/module`, for example. I can handle that eventual refactor, but I'd rather not have some pieces specified in `cli/app.go` and some specified elsewhere for now.

File: cli/module_build.go
Comment (benjirewis) [inline cli/module_build.go:416]: Spoke offline; I'd rather avoid this env-pulling functionality for now if we're just using existing auth functionality.

File: robot/impl/local_robot.go
Comment (benjirewis) [reply robot/impl/local_robot.go:1223]: I was wrong; I think `updateResources` is correct here and what you have is good.

File: robot/impl/local_robot.go
Comment (benjirewis) [inline robot/impl/local_robot.go:1235]: Extraneous todo?

File: cli/app.go
Comment (abe-winter) [reply cli/app.go:21]: irl feedback from benji: we are leaning towards using the main part id to identify the robot. (reason for this: it's what's available from viam.json).

there is some confusion here though; it's hard to get that id from the web UI. (the URL has robot id; you can get part id by going to 'debug configuration' in the ellipsis menu for the main part)

@zaporter-work we decided to punt to you -- do you have advice on 1) which ID to use for reload / restart and 2) how to communicate to users where to find it?

Comment (zaporter-work) [reply cli/app.go:21]: > main part id to identify the robot

I suspect you want to identify the robot part instead of the robot. (because you want to be able to distinguish between restarting a module on the main part vs a subpart). Please correct me if you have a reason for finding the robot instead of the part

To find the part ID, a user can use this "Part ID" button to copy the id to their clipboard:
20240418_17h07m13s_grim

Comment (zaporter-work) [reply cli/app.go:21]: We have named that dropdown the `Part status dropdown`

Comment (abe-winter) [reply cli/app.go:21]: ah this is the 'Live / Offline' pill next to the breadcrumbs on the secondary navbar -- awesome, that works

File: cli/module_build.go
Comment (abe-winter) [reply cli/module_build.go]: using viamClient auth in 6d2b7758

File: cli/module_build.go
Comment (abe-winter) [reply cli/module_build.go]: moved to app.go in 8e186a4e

File: cli/module_build.go
Comment (abe-winter) [reply cli/module_build.go]: using GetRobotApiKeys in 6d2b7758

I'm also calling GetRobotPart, which has a Part.Secrets list on it -- is there a way to use the part secret directly?

File: cli/app.go
Comment (abe-winter) [reply cli/app.go]: using part Id with UI instructions in 8e186a4e. using `Part.Fqdn` from part ID in 7dffefae5 (rather than requiring user to pass in fqdn)

is Fqdn right? Or will I need LocalFqdn sometimes

Comment (zaporter-work) [reply cli/app.go]: > Or will I need LocalFqdn sometimes

Passing the gauntlet back to @benjirewis 

I think using the Fqdn all the time will work (though perhaps LocalFqdn can improve connection performance?)

File: cli/module_build.go
Comment (zaporter-work) [reply cli/module_build.go]: I think part secrets are in a weird socally-deprecated technically-supported state on the fleet teams. If I were you, unless you write e2e tests for this, I'd avoid using the secret directly and would use the API key. But I think the secret would work

Comment (zaporter-work) [reply cli/module_build.go]: Up to you

File: cli/app.go
Comment (benjirewis) [reply cli/app.go]: I think you should use the `FQDN` and not the `LocalFQDN`. The local FQDN would, I think, only be usable if you're running the CLI on the same subnet as the machine. I realize we're probably expecting CLI users to run `viam module restart` from the same machine as the running RDK (in order to actually rebuild the module?), but we probably want to allow `viam module restart --restart-only` from a CLI user not running on the same machine. As far as performance, both Cheuk and I think the difference between using the regular and local FQDN would be fairly negligible in this case.

File: robot/impl/local_robot.go
Comment (abe-winter) [reply robot/impl/local_robot.go]: done in 0c08679ee

File: cli/module_build.go
Comment (zaporter-work) [inline cli/module_build.go:442]: `conf.Cloud` is a ptr. Should have a nil check

File: cli/module_build.go
Comment (zaporter-work) [inline cli/module_build.go:449]: Very excited about urfave/cli v3

File: cli/module_build.go
Comment (zaporter-work) [inline cli/module_build.go:486]:
```suggestion
		return errors.New("API keys list for this machine is empty. You can create one with \"viam machine api-key create\"")
```

File: robot/robot.go
Comment (zaporter-work) [inline robot/robot.go:93]:
```suggestion

        // RestartModule reloads a module as if its config changed
	RestartModule(ctx context.Context, req RestartModuleRequest) error
```

This func is exposed to all users of the go sdk I think. We probably want to add a comment

**Review by zaporter-work — APPROVED** [review]
LGTM as long as you have manually tested this. I haven't tested this.

A simple test would be nice but not trivial

File: cli/module_build.go
Comment (zaporter-work) [reply cli/module_build.go:486]: [opt]

File: robot/robot.go
Comment (zaporter-work) [reply robot/robot.go:93]: Not sure -- I might be wrong -- can ignore if I am wrong

**Comment by abe-winter** [comment]
zack re testing:
- on the CLI side, there's an e2e / mock test coming in the full reload branch (next PR)
- inside RDK, I assumed `updateResources` is already exercised -- could add something similar to [TestModuleReloading](https://github.com/viamrobotics/rdk/blob/main/module/modmanager/manager_test.go#L395) though for this specific case. your call

**Comment by abe-winter** [comment]
(and yes, have manually tested)

**Comment by zaporter-work** [comment]
>on the CLI side, there's an e2e / mock test coming in the full reload branch (next PR)
    inside RDK, I assumed updateResources is already exercised -- could add something similar to [TestModuleReloading](https://github.com/viamrobotics/rdk/blob/main/module/modmanager/manager_test.go#L395) though for this specific case. your call

I meant on the CLI side. RDK side tests LGTM

**Comment by zaporter-work** [comment]
If that is coming in the future, SGTM

File: cli/module_build.go
Comment (abe-winter) [reply cli/module_build.go:432]: done in d3b8041e

File: cli/module_build.go
Comment (abe-winter) [reply cli/module_build.go]: done in d7f5353c

File: cli/app.go
Comment (abe-winter) [reply cli/app.go]: done in 16dc8e2c

**Pull request merged**
